Sources/X509/X509BaseTypes/TimeCalculations.swift

```diff
@@ -0,0 +1,263 @@
+        if remsecs < 0 {
+            // Unchecked is safe here: we know remsecs is negative, and we know
+            // days cannot be Int64.min
+            remsecs &+= 86400
+            days &-= 1
+        }
+
+        var (qcCycles, remdays) = days.quotientAndRemainder(dividingBy: .daysPer400Years)
+        if remdays < 0 {
+            // Same justification here for unchecked arithmetic
+            remdays &+= .daysPer400Years
+            qcCycles &-= 1
+        }
+
+        // Unchecked arithmetic here is safe: the subtraction is all known values (4 - 1),
+        // and the multiplication cannot exceed the original value of remdays.
+        var cCycles = remdays / .daysPer100Years
+        if cCycles == 4 { cCycles &-= 1 }
+        remdays &-= cCycles &* .daysPer100Years
+
+        // Unchecked arithmetic here is safe: the subtraction is all known values (25 - 1),
+        // and the multiplication cannot exceed the original value of remdays.
+        var qCycles = remdays / .daysPer4Years
+        if qCycles == 25 { qCycles &-= 1 }
+        remdays &-= qCycles &* .daysPer4Years
+
+        // Unchecked arithmetic here is safe: the subtraction is all known values (4 - 1),
+        // and the multiplication cannot exceed the original value of remdays.
+        var remyears = remdays / 365
+        if remyears == 4 { remyears &-= 1 }
+        remdays &-= remyears &* 365
+
+        // Unchecked multiplication here is safe: each of these has earlier been multiplied by
+        // a much larger number and we didn't need checked math then, so we don't need it now.
+        var years = remyears + (4 &* qCycles) + (100 &* cCycles) + (400 &* qcCycles)
+
+        var months = 0
+        while Int64.daysInMonth(months) <= remdays {
+            remdays -= Int64.daysInMonth(months)
+            months += 1
+        }
+
+        // Now we normalise the months count back to starting in January.
+        // Safe to do unchecked subtraction here becuase for all numbers 10 or
+        // larger we can subtract by 12.
+        if months >= 10 {
+            months &-= 12
+            years += 1
+        }
+
+        // Normalise out the values.
+        //
+        // Safe to do unchecked math on the months as we just checked its value above.
+        // Same for remdays, the loop only terminates if the number is smaller than at most 31.
+        //
+        // Note that, unlike struct tm, we return ordinal month numbers as well as days (i.e. 1 to 12).
+        // This fits us better when working with generalizedtime and friends.
+        return (
+            year: Int(years + 2000),
+            month: Int(months &+ 3),
+            day: Int(remdays &+ 1),
+            hours: Int(remsecs / 3600),
+            minutes: Int(remsecs / 60 % 60),
+            seconds: Int(remsecs % 60)
+        )
+    }
+
+    @inlinable
+    init(timestampFromUTCDate date: (year: Int, month: Int, day: Int, hours: Int, minutes: Int, seconds: Int)) {
```

Comment:
Can we document the valid value ranges for which we guarantee not to crash and add asserts that check that the inputs are in this range?